spec/file_downloaders/s3_downloader_spec.rb

```diff
@@ -0,0 +1,72 @@
+require 'spec_helper'
+
+describe Patches::FileDownloaders::S3Downloader do
+  subject { described_class.new(bucket_name: 'jobready-s3-test', aws_config: aws_config) }
+
+  let(:aws_config) do
+    {
+      access_key_id: '123456',
+      secret_access_key: '654321',
+      region: 'ap-southeast-2'
+    }
+  end
+
+  let(:current_path) { File.expand_path(File.dirname(__FILE__)) }
+  let(:destination) { File.join(current_path, 'a.png') }
+  let(:download) { subject.download('a.png', destination) }
+
+  context 'stubbed' do
+    let(:s3) { Aws::S3::Client.new(stub_responses: true) }
+
+    before { expect(Aws::S3::Client).to receive(:new).and_return(s3) }
+
+    context 'for a successfully downloaded file' do
+      after { File.delete(destination) }
+
+      it 'saves the file' do
+        download
+        expect(File.exists?(destination)).to be_truthy
+      end
+
+      it 'returns the file path' do
+        expect(download).to eq destination
+      end
+
+      context 'credentials' do
+        let(:credentials) { Aws::Credentials.new('123456', '654321') }
+
+        before { allow(Aws::Credentials).to receive(:new).and_return(credentials) }
+
+        it 'uses the passed credentials' do
+          expect(Aws.config).to receive(:update).with(
+            region: 'ap-southeast-2',
+            credentials: credentials
+          )
+          download
+        end
+      end
+    end
+
+    context 'when there is an error downloading the file' do
+      before { s3.stub_responses(:get_object, 'NotFound') }
+
+      it 'raises an error' do
+        expect { download }.to raise_error(Aws::S3::Errors::NotFound)
+        expect(File.exists?(destination)).to be_falsey
+      end
+    end
+  end
+
+  context 'not stubbed' do
+    context 'with invalid credentials' do
+      before { WebMock.allow_net_connect! }
```

Comment:
is this actually running a test against aws servers? I don't think we should do this. We should be able to use the provided AWS library stubbing features without live connections. I'm comfortable to place faith in those libraries.